torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -533,19 +533,48 @@ RegisterOperators reg({
         prim::fork,
         [](const Node* node) {
           Code code(node->g(attr::Subgraph));
+          int n_inputs = node->inputs().size();
           JIT_ASSERT(node->blocks().size() == 0);
           JIT_ASSERT(node->hasAttribute(attr::Subgraph));
           return [=](Stack& stack) {
-            InterpreterState(code).run(stack);
-            push(stack, Future(pop(stack)));
+            // Move inputs to a separate stack
+            auto copied_stack = new Stack(
+                stack.end() - n_inputs,
+                stack.end());
+            drop(stack, n_inputs);
+
+            auto copied_state = new InterpreterState(code);
+
+            push(stack, copied_state->getFuture());
+
+            // TODO: who/how to hold the lifecycle for copied_stack/copied_state?
+            // We currently add them to callback so they will be destroyed when future completes.
+            // Suppose we have two threads: T1 forks T2 and waits on T2;
```

Comment:
You can move the stack by using a class rather than a lambda. A class with an operator() is still a valid std::function<void(void)>.
```
class InterpreterContinuation {
   InterpreterContinuation(InterpreterState state, Stack initial_stack)
  : state(std::move(state)), stack(std::move(initial_stack)) {}
  void operator()(void) {
     state.run(stack);
  }
private:
  Stack stack;
  InterpreterState state;
};
```

After the interpreter returns, whether it finished or not, the stack is no longer used. So once this Closure finishes it no longer needs Stack and it will get destroyed. If the Interpreter did not finish, then it must have created a callback on another future that would schedule the rest of the interpreter to run. That callback will capture a reference to InterpreterState, increasing its refcount. InterpreterState will destruct when there are no longer any callbacks referring to it.